[browser][websocket] Throw OperationCanceledException on connect

kjpou1 opened the pull request:
if cancel was requested before.

This fixes the test

```
        public async Task ConnectAsync_CancellationRequestedBeforeConnect_ThrowsOperationCanceledException()
        {
            using (var clientSocket = new ClientWebSocket())
            {
                var cts = new CancellationTokenSource();
                cts.Cancel();
                Task t = clientSocket.ConnectAsync(new Uri("ws://" + Guid.NewGuid().ToString("N")), cts.Token);
                await Assert.ThrowsAnyAsync<OperationCanceledException>(() => t);
            }
        }
```

https://github.com/dotnet/runtime/blob/master/src/libraries/System.Net.WebSockets.Client/tests/ConnectTest.cs#L249


Fix https://github.com/dotnet/runtime/issues/44720

Part of https://github.com/dotnet/runtime/issues/44666

Test was failing with the following error:

```
[FAIL] System.Net.WebSockets.Client.Tests.ConnectTest.ConnectAsync_CancellationRequestedBeforeConnect_ThrowsOperationCanceledException
  info: Assert.Throws() Failure
  info: Expected: typeof(System.OperationCanceledException)
  info: Actual:   typeof(System.Net.WebSockets.WebSocketException): Unable to connect to the remote server
  info: ---- System.Net.WebSockets.WebSocketException : Unable to connect to the remote server
  info: -------- System.Net.WebSockets.WebSocketException : An internal WebSocket error occurred. Please see the innerException, if present, for more details.
  info:    at System.Net.WebSockets.BrowserWebSocket.ConnectAsyncJavaScript(Uri uri, CancellationToken cancellationToken, List`1 requestedSubProtocols)
  info:    at System.Net.WebSockets.WebSocketHandle.ConnectAsync(Uri uri, CancellationToken cancellationToken, ClientWebSocketOptions options)
  info: ----- Inner Stack Trace -----
  info:    at System.Net.WebSockets.BrowserWebSocket.ConnectAsyncJavaScript(Uri uri, CancellationToken cancellationToken, List`1 requestedSubProtocols)
  info:    at System.Net.WebSockets.WebSocketHandle.ConnectAsync(Uri uri, CancellationToken cancellationToken, ClientWebSocketOptions options)
  info: ----- Inner Stack Trace -----
  info:    at System.Net.WebSockets.BrowserWebSocket.ConnectAsyncJavaScript(Uri uri, CancellationToken cancellationToken, List`1 requestedSubProtocols)

```

See https://github.com/dotnet/runtime/pull/45470 for more tests

kjpou1 committed:
[browser][websocket] Throw OperationCanceledException on connect if cancel was requested before.

stephentoub commented:
> This fixes the test

I understand that this change avoids doing extra work in case cancellation was already requested, but can you elaborate on how it fixes the test?  What was happening before this change?

kjpou1 commented:
Updated the PR comments to include the error message.

stephentoub commented:
> Test was failing with the following error:

But doesn't that suggest there's still a problem here?  What happens if cancellation is requested while the connect is in flight?  That should still produce an OperationCanceledException, but the error you pasted suggests it isn't.

lewing committed:
try to handle cancellation in connect stage

lewing commented:
The test change is a race on anything with threads so it should be changed back and and a threadless platform test added.

kjpou1 commented:
I am not finding it racey.  I will take a closer look but it seems to be working fine on my local machine.

kjpou1 committed:
Add new test for inflight connect

- Add new supported property for skipping particular tests when Browser is detected and DOM is detected.

lewing commented:
I meant in theory it might be able to fail on the host lookup before the cancellation, upon reflection that is unlikely, I'm fine either way.

kjpou1 commented:
I have made multiple tests with these modifications against a local server as well as against a public websocket test echo server with no problems.

The new test also works as expected.  The original tests tests against a server that was not found but testing against a live server where the connection actually exists and succeeds also works.

```
            using (var clientSocket = new ClientWebSocket())
            {
                var cts = new CancellationTokenSource();
                //Task t = clientSocket.ConnectAsync(new Uri("ws://" + Guid.NewGuid().ToString("N")), cts.Token);
                Task t = clientSocket.ConnectAsync(new Uri("wss://echo.websocket.org"), cts.Token);
                cts.Cancel();
                await Assert.ThrowsAnyAsync<OperationCanceledException>(() => t);
            }
```

lewing commented on src/libraries/System.Net.WebSockets.Client/tests/WebSocketHelper.cs:
```diff
@@ -13,6 +13,10 @@ public static class WebSocketHelper
     {
         private static readonly Lazy<bool> s_WebSocketSupported = new Lazy<bool>(InitWebSocketSupported);
         public static bool WebSocketsSupported { get { return s_WebSocketSupported.Value; } }
+        // On browser the JavaScript host object 'WebSocket' is only available when DOM is supported
+        // If DOM is not available then a JavaScript error is thrown
+        // --- System.Runtime.InteropServices.JavaScript.JSException : JavaScript host object 'WebSocket' not found.
+        public static bool WebSocketsSupportedInBrowser { get { return !PlatformDetection.IsBrowser || (PlatformDetection.IsBrowser && PlatformDetection.IsBrowserDomSupported); } }
```
I don't understand this new check.  WebSocketsSupported supported checks for PNSE which is what we should throw if the platform doesn't support it (and the other tests suggest that we do). If the goal is to only run this test on browser-wasm use PlatformSpecific

lewing replied:
To expand on my previous comment, if we aren't properly throwing PNSE when the JS runtime doesn't support `WebSocket` we should fix that.  If you want to skip the test on platforms other than Browser use [PlatformSpecific(TestPlatforms.Browser)]. The combination of the two should be enough?

lewing commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/WebSocketHandle.Browser.cs:
```diff
@@ -61,7 +62,8 @@ public async Task ConnectAsync(Uri uri, CancellationToken cancellationToken, Cli
                 Abort();
 
-                if (exc is WebSocketException)
+                if (exc is WebSocketException ||
+                    (exc is OperationCanceledException && cancellationToken.IsCancellationRequested))
```
I'm not sure what the optimal/expected behavior here is but this check should probably match the BrowserWebSocket case

kjpou1 replied:
This should already be resolved via commits below by yourself

lewing reviewed (commented):
The fact checks need work.

lewing committed:
first pass at throwing pnse when websocket is missing

lewing commented:
@kjpou1 I've pushed a rough pass at what I was suggesting because I realized we've discussed the complications before.

lewing commented:
And now I'm realizing the platform check was probably written that way because of the old options throwing.  If browser is the only unsupported case we simplify it a lot more.

kjpou1 commented:
Agree and agree.  The reason we did not throw if WebSocket was not found was so that we could do rudimentary tests with the javascript engines.  You are right that now that we have the tests stood up via browser we can now do it this way.

kjpou1 commented:
@lewing ping.  Can you review this again?

kjpou1 commented:
/cc @marek-safar @stephentoub

marek-safar commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/WebSocketHandle.Browser.cs:
```diff
@@ -14,7 +14,11 @@ internal sealed class WebSocketHandle
-        public static ClientWebSocketOptions CreateDefaultOptions() => new ClientWebSocketOptions();
+        public static ClientWebSocketOptions CreateDefaultOptions()
+        {
+            BrowserWebSocket.ThrowIfPlatformNotSupported();
```
Is this the right place for the throw? Which calls are you trying to protect?

kjpou1 replied:
I would say yes and not protecting any calls that I know of.  It allows for checking in the tests if we are able to run the tests for browser or not.

marek-safar replied:
Then why not to do the checks in the tests and exclude the tests on configurations which don't support websockets?

kjpou1 replied:
The tests check for PlatformNotSupported

kjpou1 replied:
Also, the original code did the checks in the tests and it was suggested to do it this way.  Whichever way is ok with me just let me know.

lewing replied:
I don't think it makes any sense at this point to have WebSocketsSupported be true on v8 and then add another condition for when it is really true on browser.  If browser is the only platform that doesn't support WebSockets can't the condition can just become PlatformDetection.IsNotBrowser || PlatformDetection.IsBrowserDomSupported.  If there isn't a easy way to express that in attributes we can either stick it in InitWebSocketSupported or add another PlatformDetection value, something like IsClientNetworkingSupported which would also cover the outgoing http case.

lewing replied:
WebSocketsSupported is used other places so I moved the check to the tests only, if we ever intend to officially support wasm-browser on !browser we'll have more places to fix than just here.

kjpou1 replied:
Thanks Larry

marek-safar commented on src/libraries/System.Net.WebSockets.Client/tests/WebSocketHelper.cs:
```diff
@@ -13,7 +13,6 @@ public static class WebSocketHelper
     {
         private static readonly Lazy<bool> s_WebSocketSupported = new Lazy<bool>(InitWebSocketSupported);
         public static bool WebSocketsSupported { get { return s_WebSocketSupported.Value; } }
-
```
Revert this change

kjpou1 replied:
addressed

kjpou1 committed:
Merge branch 'master' of https://github.com/dotnet/runtime into wasm-websockets-connect-test

kjpou1 committed:
Address review comment

lewing committed:
Make the platform check explicit

lewing committed:
Revert CreateDefaultOptions change

marek-safar commented on src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/WebSocketHandle.Browser.cs:
```diff
@@ -61,11 +62,14 @@ public async Task ConnectAsync(Uri uri, CancellationToken cancellationToken, Cli
-                    throw;
+                switch (exc) {
+                    case WebSocketException:
+                    case PlatformNotSupportedException:
```
Is this case still needed?

kjpou1 replied:
Addressed in recent commit

kjpou1 committed:
Address review comment

marek-safar commented:
Looks good to me but will wait on @lewing to approve
